src/testing/cmds/harness/test-runner/main.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/google/shlex"
+	"github.com/pachyderm/pachyderm/v2/src/internal/backoff"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+)
+
+func main() {
+	log.InitPachctlLogger()
+	ctx := pctx.Background("test-runner")
+	tags := flag.String("tags", "", "Tags to run, for example k8s. Tests without this flag will not be selected.")
+	fileName := flag.String("file", "tests_to_run.csv", "csv file containing the list of test names to run generated by the collector.")
+	gotestsumArgsRaw := flag.String("gotestsum-args", "", "Additional arguments to pass to the gotestsum portion of the test command.")
+	gotestArgsRaw := flag.String("gotest-args", "", "Additional arguments to pass to the 'go test' portion of the test command.")
+	shard := flag.Int("shard", 0, "0 indexed current runner index that we are on.")
+	totalShards := flag.Int("total-shards", 1, "Total number of runners that we are sharding over.")
+	threadPool := flag.Int("threads", 1, "Number of tests to execute concurrently.")
+	flag.Parse()
+	gotestsumArgs, err := shlex.Split(*gotestsumArgsRaw)
+	if err != nil {
+		log.Error(ctx, "Error parsing gotestsumArgs", zap.Error(err))
+		os.Exit(1)
+	}
+	gotestArgs, err := shlex.Split(*gotestArgsRaw)
+	if err != nil {
+		log.Error(ctx, "Error parsing gotestArgs", zap.Error(err))
+		os.Exit(1)
+	}
+
+	err = run(ctx,
+		*tags,
+		*fileName,
+		gotestsumArgs,
+		gotestArgs,
+		*shard,
+		*totalShards,
+		*threadPool,
+	)
+	if err != nil {
+		log.Error(ctx, "Error running tests", zap.Error(err))
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+func run(ctx context.Context, tags string, fileName string, gotestsumArgs []string, gotestArgs []string, shard int, totalShards int, threadPool int) error {
+	tests, err := readTests(ctx, fileName)
+	if err != nil {
+		return errors.Wrapf(err, "reading file %v", fileName)
+	}
+	slices.Sort(tests) // sort so we shard them from a pre-determined order
```

Comment:
I don't think this is strictly necessary.  Hash + modular arithmetic to number of shards will yield the same result.  I like xxhash3 for this; `xxh3.Hash(testName) % totalShards`.   (We already use xxh3 for computing starlark hashes of go objects.)

This is O(1) instead of O(n log n) over the total number of tests, and even if more tests are added, tests stay in the shard they were originally assigned (unless totalShards changes, of course).

Reply:
BTW, if you want stability as the shard count changes, that's consistent hashing.  We have an algorithm for that in internal, but the gist is that you map each test to a value on a unit circle (between 0 radians and 2*pi/math.MaxUint64), then divide the circle into totalShards equally-sized fractions, then assign the test to a shard based on which fraction the point representing the test falls into.  This means that as you increase the number of test runners, most of the tests stay in the same place.  Sorting accomplishes a similar goal, but changes shard position when tests are added or removed.

Reply:
Neat! We only need per-run consistency here so multiple shards don't take the same test and tests don't get missed in the current run, but good to know.

Reply:
Update, I really like the hashing strategy but it is not a fully even distribution, so it ends up being a lot slower. I reverted it in favor of the most even test distribution. 

The problem I'm solving here is that circle-ci doesn't have a way for me to distribute tests to a runner that is free when it becomes free. I only have the total number of nodes, and which node I am currently on. For integration tests especially distributing 30 tests to one runner and 34 tests to another means the whole job will be significantly slower than running 32 on both since the bottle neck is minikube install time.

I considered using the pachops cluster to coordinate and do test distribution in real time to more efficiently use runners, but I suspect that the complexity and additional dependency would bite us long term, so I'm deferring for now.

Reply:
that makes sense.  keep what you have!